Add wishbone packet buffers

lmbollen opened the pull request:
This pull requests adds the components important for our ethernet stack, these components are written in Axi4Stream, but will in the future probably be replaced with the currently experimental `PacketStream` due to lower complexity overhead.
 
1. A wishbone to axi transmitter `wbToAxiTx`:
 A component with a wishbone interface that produces an AXI transaction.
 Writing to address 1 will produce a transaction with _tlast set and all keep enables low. 
This allows us to write the entire packet to address 0 and then write to address 1 to signal the end of the packet without having to keep track of the last byte in software.

2. A wishbone accessible AXI receive buffer `wbAxisRxBuffer`:
This component contains a blockram to which AXI transactions will be written. Except for the last transactions, these transactions can *not* contain null bytes, since they will be written to the buffer as words. The memory layout is as follows:
```
0 .. 4 * (depth - 1) : The packet buffer
(4 * depth)            :  Packet length in bytes
(4 * (depth + 1)     : Status register (packet Complete and buffer full)
```
To read a packet we check the status register, if it contains data (as indicated by packet complete, buffer full or both), we read can read the packet from the buffer.

To clear a packet, we set the packet length register and status register to 0 in that order.
This buffer can hold a single packet and will stall incoming packets when it contains a complete packet.

3. A axi 1 to n scaling and packing circuit. `axisFromByteStream`
We receive the Ethernet packet as a 1 byte wide stream, but want to write it to the buffer as n bytes wide without gaps. Therefore we have to perform upscaling and packing.

4. An axi n to 1 scaling circuit. `axisToByteStream`
We produce a stream of n bytes wide, but the GMII interface we produce is 1 byte wide, therefore we have to downscale the stream to a single byte wide. 

5. Axi packing circuit `axiPacking`
For arbitrary packing without compromising on throughput. We don't currently test this.

6. Axi packet fifo `axiStreamPacketFifo`
A buffer that first consumes a complete packet and then produces it as a packet without gaps when complete (or the buffer is full)

Furthermore it introduces various utility functions for manipulating axi stream transactions

christiaanb commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+    undefUser = deepErrorX "axisFromByteStream: _tuser undefined"
+    -- Try to append the incoming axi to the stored axi
+    combinedAxi = axiUserMap (\(v1, v2) -> v1 :< v2) <$> concatAxi axiStored axiIn
+    inpValid = isJust axiIn && isJust combinedAxi
```
perhaps `inpPass`?

christiaanb commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+    -- Try to append the incoming axi to the stored axi
+    combinedAxi = axiUserMap (\(v1, v2) -> v1 :< v2) <$> concatAxi axiStored axiIn
+    inpValid = isJust axiIn && isJust combinedAxi
+    inpInvalid = isJust axiIn && isNothing combinedAxi
```
Perhaps `inpBlocked`?

christiaanb commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+      (maybe (repeat undefUser) _tuser axiPreShift)
+
+    (newData, newKeeps, newStrobes, newUsers) = unzip4 $
+      (maybe (0, False, False, undefUser) (\(d, s, u) -> (d, True, s, u))) <$> shiftPackVec axiBytes
```
This `shiftPackVec` seems expensive, when we know that only the `axiIn`  should contain a `tkeep` that's deasserted.

I guess it's acceptable for now. But do create an issue on the  issue tracker that this function should be made more resource efficient.

lmbollen commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+  aValid = aKeepBytes || lastA
+  bValid = bKeepBytes
+  (aKeepBytes, bKeepBytes) = (or keepA, or keepB)
+  lastA = _tlast axi && not bKeepBytes
+  lastB = _tlast axi && bKeepBytes
```
The two resulting AXI transactions belong to the same stream, setting `_tlast` in both outputs would result in a second empty packet.

lmbollen commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+-- bits is set. If _tlast is set, the _tlast of the first output is set if the second
+-- output is `Nothing`. If the second output is `Just`, the _tlast of the first output
+-- is `False` and the _tlast of the second output is `True`.
+splitAxi ::
```
Then we can't test them.

lmbollen commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+   where
+
+    -- Axi Right
+    validRight = maybe False (and . _tkeep) outputBuffer || maybe False _tlast outputBuffer || inpInvalid
```
Otherwise a malformed packet would lock up the bus.

lmbollen commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+      (True, _) -> (wbData, wbHandshake && not readingFifo, wbHandshake && readingFifo)
+      (False, _) -> (deepErrorX "undefined", False, False)
+
+    axisReady = packetError || not (packetComplete || bufferFull)
```
We can currently only store a single packet

lmbollen commented on bittide/src/Bittide/Axi4.hs:
```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+-- cleared.
+--
+-- The external status clear signals clear on True, set to (False, False) if not used.
+wbAxisRxBuffer ::
```
Personally I think we should switch over to a simpler protocol. AXI4Stream is too versatile to have generic components.

martijnbastiaan commented on bittide-instances/tests/Wishbone/Axi.hs:
```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# OPTIONS_GHC -fplugin=Protocols.Plugin #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=0 #-}
```
Don't do this please. If there is a bug in the type checker this can cause it to go into an infinite loop and now you suddenly have a problem whose origin is hard to detect.

martijnbastiaan commented:
@lmbollen Could you look at the failing test in `Axi4.hs`? Uncomment the block and then:

```
cd bittide
cabal run unittests -- -j12 --hide-successes -p "axiStreamFromByteStream produces dense streams"
```

martijnbastiaan commented:
Almost there.. According to [my document :sparkles:](https://docs.google.com/spreadsheets/d/1iVRMWumZbVGUo3hlURAptXUnKqouaBs7HjWeTwPMVmU/edit?gid=0#gid=0), we're still missing:

- [ ] Direct tests for `wbToAxiTx`
- [ ] Direct tests for `packAxi4Stream`
- [ ] Direct tests for `unInterruptedAxi4Packets`
- [ ] A bugfix for the now commented `axiStreamFromByteStream produces dense streams`

lmbollen commented:
@martijnbastiaan I'd like to discuss if and how we would test `wbToAxiTx`

For now I've:
* added verification functions for properties we extracted from the spec
* Fixed the up and downscaling circuits
* Added tests for the properties and other functions

martijnbastiaan commented on bittide-instances/tests/Wishbone/Axi.hs:
```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# OPTIONS_GHC -fplugin=Protocols.Plugin #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=0 #-}
```
_Bro_

lmbollen replied:
https://github.com/bittide/bittide-hardware/pull/589/commits/968b2c43c888b4618c770574dfc862e745a4bda8 :eyes:

martijnbastiaan reviewed (approved):
Let's make an issue for the missing test and let's call it a day!
